refactor(examples): HepMCWriter and Reader as algorithm

FabianKlimpel opened the pull request:
This PR modifies the `HepMCWriter` and `HepMCReader` such that they can be used as algorithms in the `Sequencer`.

msmk0 commented on Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Reader.hpp:
```diff
@@ -1,21 +1,31 @@
+struct HepMC3ReaderAscii : public IReader {
+  struct Config {
+    // The input directory
+    std::string inputDir = "";
```
```suggestion
    std::string inputDir;
```
Nitpicking: default initialization already provides an empty string.

FabianKlimpel replied:
done

msmk0 commented on Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Reader.hpp:
```diff
@@ -1,21 +1,31 @@
+    // The input directory
+    std::string inputDir = "";
+    // The stem of input file names
+    std::string inputStemFileName = "";
```
```suggestion
    std::string inputStem;
```
Stem is sufficient and consistent with e.g. the csv reader/writer.

FabianKlimpel replied:
Consistent yes, but not sufficient imo. Changed

msmk0 replied:
Hmm, yes. One could argue about that. Thanks for changing it anyways to be consistent.

msmk0 commented on Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Reader.hpp:
```diff
@@ -1,21 +1,31 @@
+    // The stem of input file names
+    std::string inputStemFileName = "";
+    // The output collection
+    std::string outputCollection = "";
```
Can you name this a bit more specific. Maybe `outputEvent` similar to the existing names collection names like `outputSimHits` or `outputParticles`.

FabianKlimpel replied:
outputEvents it is

msmk0 commented on Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Writer.hpp:
```diff
@@ -1,33 +1,51 @@
+    : public WriterT<std::vector<std::shared_ptr<HepMC3::GenEvent>>> {
+  struct Config {
+    // The output directory
+    std::string outputDir = "";
```
```suggestion
    std::string outputDir;
```
Nitpicking: see above.

FabianKlimpel replied:
done

msmk0 commented on Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Writer.hpp:
```diff
@@ -1,33 +1,51 @@
+    // The output directory
+    std::string outputDir = "";
+    // The stem of output file names
+    std::string outputStemFileName = "";
```
```suggestion
    std::string outputStem;
```
Stem is already sufficient. This is also consistent with the other configurations, e.g. the csv reader/writer.

FabianKlimpel replied:
done

msmk0 commented on Examples/Io/HepMC3/src/HepMC3Reader.cpp:
```diff
@@ -17,3 +20,47 @@ bool ActsExamples::HepMC3ReaderAscii::readEvent(
+
+  ctx.eventStore.add(m_cfg.outputCollection, std::move(events));
+
+  return reader.failed() ? ActsExamples::ProcessCode::ABORT
+                         : ActsExamples::ProcessCode::SUCCESS;
```
If the reading has a failure the output collection might be missing something. The failure check should therefore come before the collection is written to the whiteboard.

FabianKlimpel replied:
fixed

msmk0 commented on Examples/Io/HepMC3/src/HepMC3Writer.cpp:
```diff
@@ -1,20 +1,33 @@
+    writer.write_event(*event);
+
+  writer.close();
+  return writer.failed() ? ActsExamples::ProcessCode::ABORT
```
The failure check should occur after each write. I do not not whether i/o error persist in the writer, but it is more robust if you detect errors as soon as they can occur.

FabianKlimpel replied:
fixed

msmk0 reviewed (changes requested):
@FabianKlimpel Thank you splitting this up. I added some specific comments in the code. I also have a few general questions/suggestions:

- Please add a separate header file that provides a typedef for the read/written event type. Something like `HepMC3Event` is much easier to read compared to `std::vector<std::shared_ptr<HepMC3::GenEvent>>` and it simplifies life for users. Similar to the various EDM data containers e.g. `SimHitsContainer`.
- Is is necessary to store the `HepMC3::GenEvent` by shared pointer? Would it be possible to store it by value instead, i.e. `std::vector<HepMC3::GenEvent>`? Since the collection is moved to the whiteboard and can not be modified, there is no problem of shared ownership. Its lifetime is fixed until the event processing is done anyways. Is this a requirement from HepMC?
- The reader still contain some methods (`readEvent` and `status`) that are not needed anymore to satisfy the algorithm interface. They should be removed.
- For consistency with the other I/O algorithms, I would suggest to rename the algorithms to `AsciiHepMC3{Reader,Writer}` or `HepMC3Ascii{Reader,Writer}`.
- The HepMC3 example contains both the new sequencer-based code as well as the manual reading using the HepMC3 directly. My suggestion would be to use only the sequencer-based code and remove the rest.

FabianKlimpel commented on Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Reader.hpp:
```diff
@@ -1,21 +1,31 @@
 #pragma once
 
+#include "ActsExamples/Framework/IReader.hpp"
+#include <Acts/Utilities/Logger.hpp>
```
done

FabianKlimpel commented on Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Reader.hpp:
```diff
@@ -1,21 +1,31 @@
 /// HepMC3 event reader.
-struct HepMC3ReaderAscii {
- public:
+struct HepMC3ReaderAscii : public IReader {
```
done

FabianKlimpel commented on Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Writer.hpp:
```diff
@@ -1,33 +1,51 @@
-  /// always true.
-  bool writeEvent(HepMC3::WriterAscii& writer,
-                  std::shared_ptr<HepMC3::GenEvent> event);
+struct HepMC3WriterAscii final
```
done

FabianKlimpel commented on Examples/Io/HepMC3/src/HepMC3Writer.cpp:
```diff
@@ -1,20 +1,33 @@
+    const std::vector<std::shared_ptr<HepMC3::GenEvent>>& events) {
+  auto path = perEventFilepath(
+      m_cfg.outputDir, m_cfg.outputStemFileName + ".hepmc3", ctx.eventNumber);
+  std::cout << "Path: " << path << std::endl;
```
fixed

FabianKlimpel commented:
> @FabianKlimpel Thank you splitting this up. I added some specific comments in the code. I also have a few general questions/suggestions:
> 
> * Please add a separate header file that provides a typedef for the read/written event type. Something like `HepMC3Event` is much easier to read compared to `std::vector<std::shared_ptr<HepMC3::GenEvent>>` and it simplifies life for users. Similar to the various EDM data containers e.g. `SimHitsContainer`.
> * Is is necessary to store the `HepMC3::GenEvent` by shared pointer? Would it be possible to store it by value instead, i.e. `std::vector<HepMC3::GenEvent>`? Since the collection is moved to the whiteboard and can not be modified, there is no problem of shared ownership. Its lifetime is fixed until the event processing is done anyways. Is this a requirement from HepMC?
> * The reader still contain some methods (`readEvent` and `status`) that are not needed anymore to satisfy the algorithm interface. They should be removed.
> * For consistency with the other I/O algorithms, I would suggest to rename the algorithms to `AsciiHepMC3{Reader,Writer}` or `HepMC3Ascii{Reader,Writer}`.
> * The HepMC3 example contains both the new sequencer-based code as well as the manual reading using the HepMC3 directly. My suggestion would be to use only the sequencer-based code and remove the rest.

Let me reply in a different order:
I removed `std::shared_ptr<..>` and pass the references instead. The name became shorter and I would like to avoid the typedef in another file as it has the opposite effect: It gets more frustrating to look it up. The class names were changed. The reader was just implemented due to symmetry reasons, mentioned in #516. The `HepMC3Example` is still running without the sequencer and therefore the `readEvent` and `status` are still needed for this file. This will require a clean-up of this example that should be covered by a different PR.

msmk0 commented on Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Options.hpp:
```diff
@@ -0,0 +1,44 @@
+
+namespace Options {
+
+/// @brief HepMC3 specific writer options
```
```suggestion
/// Add HepMC3 specific writer options.
```
Nitpicking: `@brief` is not required.

FabianKlimpel replied:
done

msmk0 commented on Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Options.hpp:
```diff
@@ -0,0 +1,44 @@
+/// @returns a Config object for the HepMC3WriterAscii
+HepMC3AsciiWriter::Config readHepMC3WriterOptions(const Variables& variables);
+
+/// @brief HepMC3 specific reader options
```
```suggestion
/// Add HepMC3 specific reader options.
```
Nitpicking: see above.

FabianKlimpel replied:
done

msmk0 commented on Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Writer.hpp:
```diff
@@ -1,33 +1,51 @@
-  /// @param writer writer of run files
-  /// @return boolean status indicator
-  bool status(HepMC3::WriterAscii& writer);
+  /// @brief Constructor
```
```suggestion
  /// Construct the writer.
```
Nitpicking: `@brief` is not required.

FabianKlimpel replied:
done

msmk0 commented on Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Writer.hpp:
```diff
@@ -1,33 +1,51 @@
+  /// @param [in] lvl The level of the logger
+  HepMC3AsciiWriter(const Config&& cfg, Acts::Logging::Level lvl);
+
+  /// @brief Writing method
```
```suggestion
  /// Writing events to file.
```
Nitpicking: see above.

FabianKlimpel replied:
done

msmk0 reviewed (changes requested):
@FabianKlimpel thanks for adopting the fixes. This looks good and I have only some minor stylistic comments. I will leave it up to you whether you want to incorporate them or leave it as it is.

As a general comment: I would recommend not to add the HepMC3 specific input/output dir options and instead to use the general `input-dir` and `output-dir` option instead to have a consistent interface between different tools.

For my own curiosity: would it be possible to merge the multiple events read from file into a single `GenEvent`?

> I removed std::shared_ptr<..> and pass the references instead. The name became shorter and I would like to avoid the typedef in another file as it has the opposite effect: It gets more frustrating to look it up.

Ok.

> The HepMC3Example is still running without the sequencer and therefore the readEvent and status are still needed for this file. This will require a clean-up of this example that should be covered by a different PR.

Ok.

FabianKlimpel commented:
> As a general comment: I would recommend not to add the HepMC3 specific input/output dir options and instead to use the general `input-dir` and `output-dir` option instead to have a consistent interface between different tools.

Is replacing the `hepmc-in/output-directory` by `in/ouput-dir` fine or is it problematic if the same tag is used multiple times?

> For my own curiosity: would it be possible to merge the multiple events read from file into a single `GenEvent`?

The event is just a collection of vertices, so this is perfectly fine.

msmk0 commented:
> Is replacing the `hepmc-in/output-directory` by `in/ouput-dir` fine or is it problematic if the same tag is used multiple times?

This is done almost everywhere else in the code. In that case, you would not add `input-dir` via the HepMC3 options but via the generic options. When reading the config you can check if it is set and fall-back to an empty directory if not. See how this is done e.g. for the Csv reader options.

> 
> > For my own curiosity: would it be possible to merge the multiple events read from file into a single `GenEvent`?
> 
> The event is just a collection of vertices, so this is perfectly fine.
